File: sdk/java/src/main/java/tusk/license/License.java
package tusk.license;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * TuskLang SDK License Validation Module
 * Enterprise-grade license validation for Java SDK
 */
public class TuskLicense {
    private static final Logger logger = LoggerFactory.getLogger(TuskLicense.class);
    private static final String DEFAULT_SERVER_URL = "https://api.tusklang.org/v1/license";
    private static final ObjectMapper objectMapper = new ObjectMapper();
    
    private final String licenseKey;
    private final String apiKey;
    private final String sessionId;
    private final Map<String, LicenseCacheEntry> licenseCache;
    private final List<ValidationAttempt> validationHistory;
    private final List<ExpirationWarning> expirationWarnings;
    private final HttpClient httpClient;
    private final Path cacheDir;
    private final Path cacheFile;
    private OfflineCacheData offlineCache;
    
    public TuskLicense(String licenseKey, String apiKey) {
        this(licenseKey, apiKey, null);
    }
    
    public TuskLicense(String licenseKey, String apiKey, String cacheDir) {
        this.licenseKey = licenseKey;
        this.apiKey = apiKey;
        this.sessionId = UUID.randomUUID().toString();
        this.licenseCache = new ConcurrentHashMap<>();
        this.validationHistory = Collections.synchronizedList(new ArrayList<>());
        this.expirationWarnings = Collections.synchronizedList(new ArrayList<>());
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        
        // Set up offline cache directory
        if (cacheDir != null) {
            this.cacheDir = Paths.get(cacheDir);
        } else {
            String userHome = System.getProperty("user.home");
            this.cacheDir = Paths.get(userHome, ".tusk", "license_cache");
        }
        
        try {
            Files.createDirectories(this.cacheDir);
        } catch (IOException e) {
            logger.error("Failed to create cache directory", e);
        }
        
        // Generate cache file name based on license key hash
        try {
            MessageDigest md5 = MessageDigest.getInstance("MD5");
            byte[] hash = md5.digest(licenseKey.getBytes(StandardCharsets.UTF_8));
            String hashString = bytesToHex(hash);
            this.cacheFile = this.cacheDir.resolve(hashString + ".cache");
        } catch (Exception e) {
            throw new RuntimeException("Failed to initialize cache file", e);
        }
        
        // Load offline cache if exists
        loadOfflineCache();
    }
    
    public ValidationResult validateLicenseKey() {
        try {
            if (licenseKey == null || licenseKey.length() < 32) {
                return new ValidationResult(false, "Invalid license key format", null);
            }
            
            if (!licenseKey.startsWith("TUSK-")) {
                return new ValidationResult(false, "Invalid license key prefix", null);
            }
            
            MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
            byte[] hash = sha256.digest(licenseKey.getBytes(StandardCharsets.UTF_8));
            String checksum = bytesToHex(hash);
            
            if (!checksum.startsWith("tusk")) {
                return new ValidationResult(false, "Invalid license key checksum", null);
            }
            
            return new ValidationResult(true, null, checksum);
            
        } catch (Exception e) {
            return new ValidationResult(false, e.getMessage(), null);
        }
    }
    
    public Map<String, Object> verifyLicenseServer() {
        return verifyLicenseServer(DEFAULT_SERVER_URL);
    }
    
    public Map<String, Object> verifyLicenseServer(String serverUrl) {
        try {
            long timestamp = Instant.now().getEpochSecond();
            
            ObjectNode data = objectMapper.createObjectNode();
            data.put("license_key", licenseKey);
            data.put("session_id", sessionId);
            data.put("timestamp", timestamp);
            
            // Generate signature
            String dataJson = objectMapper.writeValueAsString(data);
            String signature = generateHmacSignature(dataJson, apiKey);
            data.put("signature", signature);
            
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(serverUrl))
                    .header("Authorization", "Bearer " + apiKey)
                    .header("Content-Type", "application/json")
                    .header("User-Agent", "TuskLang-Java-SDK/1.0.0")
                    .timeout(Duration.ofSeconds(10))
                    .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(data)))
                    .build();
            
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            
            if (response.statusCode() == 200) {
                Map<String, Object> result = objectMapper.readValue(response.body(), Map.class);
                
                // Update in-memory cache
                licenseCache.put(licenseKey, new LicenseCacheEntry(
                        result,
                        timestamp,
                        timestamp + 3600 // Cache for 1 hour
                ));
                
                // Update offline cache
                saveOfflineCache(result);
                
                return result;
            } else {
                logger.warn("Server returned error: {}", response.statusCode());
                return fallbackToOfflineCache("Server error: " + response.statusCode());
            }
            
        } catch (IOException | InterruptedException e) {
            logger.warn("Network error during license validation: {}", e.getMessage());
            return fallbackToOfflineCache("Network error: " + e.getMessage());
        } catch (Exception e) {
            logger.error("Unexpected error during license validation", e);
            return fallbackToOfflineCache(e.getMessage());
        }
    }
    
    public ExpirationResult checkLicenseExpiration() {
        try {
            String[] parts = licenseKey.split("-");
            if (parts.length < 4) {
                return new ExpirationResult(true, "Invalid license key format", null, null, null, null);
            }
            
            String expirationStr = parts[parts.length - 1];
            long expirationTimestamp = Long.parseLong(expirationStr, 16);
            Instant expirationDate = Instant.ofEpochSecond(expirationTimestamp);
            Instant currentDate = Instant.now();
            
            if (expirationDate.isBefore(currentDate)) {
                long daysOverdue = Duration.between(expirationDate, currentDate).toDays();
                return new ExpirationResult(
                        true,
                        null,
                        expirationDate.atZone(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT),
                        null,
                        daysOverdue,
                        null
                );
            }
            
            long daysRemaining = Duration.between(currentDate, expirationDate).toDays();
            
            if (daysRemaining <= 30) {
                expirationWarnings.add(new ExpirationWarning(
                        Instant.now().getEpochSecond(),
                        daysRemaining
                ));
            }
            
            return new ExpirationResult(
                    false,
                    null,
                    expirationDate.atZone(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT),
                    daysRemaining,
                    null,
                    daysRemaining <= 30
            );
            
        } catch (Exception e) {
            return new ExpirationResult(true, e.getMessage(), null, null, null, null);
        }
    }
    
    public PermissionResult validateLicensePermissions(String feature) {
        try {
            // Check cached license data
            LicenseCacheEntry cacheEntry = licenseCache.get(licenseKey);
            if (cacheEntry != null && Instant.now().getEpochSecond() < cacheEntry.expires) {
                Map<String, Object> licenseData = cacheEntry.data;
                if (licenseData.containsKey("features")) {
                    List<String> features = (List<String>) licenseData.get("features");
                    if (features.contains(feature)) {
                        return new PermissionResult(true, feature, null);
                    } else {
                        return new PermissionResult(false, feature, "Feature not licensed");
                    }
                }
            }
            
            // Fallback to basic validation
            switch (feature) {
                case "basic":
                case "core":
                case "standard":
                    return new PermissionResult(true, feature, null);
                case "premium":
                case "enterprise":
                    if (licenseKey.toUpperCase().contains("PREMIUM") || 
                        licenseKey.toUpperCase().contains("ENTERPRISE")) {
                        return new PermissionResult(true, feature, null);
                    } else {
                        return new PermissionResult(false, feature, "Premium license required");
                    }
                default:
                    return new PermissionResult(false, feature, "Unknown feature");
            }
            
        } catch (Exception e) {
            return new PermissionResult(false, feature, e.getMessage());
        }
    }
    
    public LicenseInfo getLicenseInfo() {
        ValidationResult validationResult = validateLicenseKey();
        ExpirationResult expirationResult = checkLicenseExpiration();
        
        LicenseInfo info = new LicenseInfo();
        info.licenseKey = licenseKey.substring(0, Math.min(8, licenseKey.length())) + "..." + 
                         licenseKey.substring(Math.max(0, licenseKey.length() - 4));
        info.sessionId = sessionId;
        info.validation = validationResult;
        info.expiration = expirationResult;
        info.cacheStatus = licenseCache.containsKey(licenseKey) ? "cached" : "not_cached";
        info.validationCount = validationHistory.size();
        info.warnings = expirationWarnings.size();
        
        LicenseCacheEntry cacheEntry = licenseCache.get(licenseKey);
        if (cacheEntry != null) {
            info.cachedData = cacheEntry.data;
            info.cacheAge = Instant.now().getEpochSecond() - cacheEntry.timestamp;
        }
        
        return info;
    }
    
    public void logValidationAttempt(boolean success, String details) {
        validationHistory.add(new ValidationAttempt(
                Instant.now().getEpochSecond(),
                success,
                details,
                sessionId
        ));
    }
    
    private void loadOfflineCache() {
        try {
            if (Files.exists(cacheFile)) {
                String json = Files.readString(cacheFile);
                offlineCache = objectMapper.readValue(json, OfflineCacheData.class);
                
                // Verify the cache is for the correct license key
                MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
                byte[] hash = sha256.digest(licenseKey.getBytes(StandardCharsets.UTF_8));
                String keyHash = bytesToHex(hash);
                
                if (offlineCache.licenseKeyHash.equals(keyHash)) {
                    logger.info("Loaded offline license cache");
                } else {
                    logger.warn("Offline cache key mismatch");
                    offlineCache = null;
                }
            }
        } catch (Exception e) {
            logger.error("Failed to load offline cache", e);
            offlineCache = null;
        }
    }
    
    private void saveOfflineCache(Map<String, Object> licenseData) {
        try {
            MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
            byte[] hash = sha256.digest(licenseKey.getBytes(StandardCharsets.UTF_8));
            String keyHash = bytesToHex(hash);
            
            OfflineCacheData cacheData = new OfflineCacheData();
            cacheData.licenseKeyHash = keyHash;
            cacheData.licenseData = licenseData;
            cacheData.timestamp = Instant.now().getEpochSecond();
            cacheData.expiration = checkLicenseExpiration();
            
            String json = objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(cacheData);
            Files.writeString(cacheFile, json);
            offlineCache = cacheData;
            logger.info("Saved license data to offline cache");
            
        } catch (Exception e) {
            logger.error("Failed to save offline cache", e);
        }
    }
    
    private Map<String, Object> fallbackToOfflineCache(String errorMsg) {
        if (offlineCache != null && offlineCache.licenseData != null) {
            long cacheAge = Instant.now().getEpochSecond() - offlineCache.timestamp;
            double cacheAgeDays = cacheAge / 86400.0;
            
            // Check if cached license is not expired
            if (!offlineCache.expiration.expired) {
                logger.warn("Using offline license cache (age: {:.1f} days)", cacheAgeDays);
                Map<String, Object> result = new HashMap<>(offlineCache.licenseData);
                result.put("offline_mode", true);
                result.put("cache_age_days", cacheAgeDays);
                result.put("warning", "Operating in offline mode due to: " + errorMsg);
                return result;
            } else {
                Map<String, Object> result = new HashMap<>();
                result.put("valid", false);
                result.put("error", "License expired and server unreachable: " + errorMsg);
                result.put("offline_cache_expired", true);
                return result;
            }
        } else {
            Map<String, Object> result = new HashMap<>();
            result.put("valid", false);
            result.put("error", "No offline cache available: " + errorMsg);
            result.put("offline_cache_missing", true);
            return result;
        }
    }
    
    private String generateHmacSignature(String data, String key) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKeySpec = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(secretKeySpec);
        byte[] hmacBytes = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(hmacBytes);
    }
    
    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }
    
    // Data classes
    public static class ValidationResult {
        public final boolean valid;
        public final String error;
        public final String checksum;
        
        public ValidationResult(boolean valid, String error, String checksum) {
            this.valid = valid;
            this.error = error;
            this.checksum = checksum;
        }
    }
    
    public static class ExpirationResult {
        public final boolean expired;
        public final String error;
        public final String expirationDate;
        public final Long daysRemaining;
        public final Long daysOverdue;
        public final Boolean warning;
        
        public ExpirationResult(boolean expired, String error, String expirationDate,
                              Long daysRemaining, Long daysOverdue, Boolean warning) {
            this.expired = expired;
            this.error = error;
            this.expirationDate = expirationDate;
            this.daysRemaining = daysRemaining;
            this.daysOverdue = daysOverdue;
            this.warning = warning;
        }
    }
    
    public static class PermissionResult {
        public final boolean allowed;
        public final String feature;
        public final String error;
        
        public PermissionResult(boolean allowed, String feature, String error) {
            this.allowed = allowed;
            this.feature = feature;
            this.error = error;
        }
    }
    
    public static class LicenseInfo {
        public String licenseKey;
        public String sessionId;
        public ValidationResult validation;
        public ExpirationResult expiration;
        public String cacheStatus;
        public int validationCount;
        public int warnings;
        public Map<String, Object> cachedData;
        public Long cacheAge;
    }
    
    private static class LicenseCacheEntry {
        public final Map<String, Object> data;
        public final long timestamp;
        public final long expires;
        
        public LicenseCacheEntry(Map<String, Object> data, long timestamp, long expires) {
            this.data = data;
            this.timestamp = timestamp;
            this.expires = expires;
        }
    }
    
    private static class ValidationAttempt {
        public final long timestamp;
        public final boolean success;
        public final String details;
        public final String sessionId;
        
        public ValidationAttempt(long timestamp, boolean success, String details, String sessionId) {
            this.timestamp = timestamp;
            this.success = success;
            this.details = details;
            this.sessionId = sessionId;
        }
    }
    
    private static class ExpirationWarning {
        public final long timestamp;
        public final long daysRemaining;
        
        public ExpirationWarning(long timestamp, long daysRemaining) {
            this.timestamp = timestamp;
            this.daysRemaining = daysRemaining;
        }
    }
    
    private static class OfflineCacheData {
        public String licenseKeyHash;
        public Map<String, Object> licenseData;
        public long timestamp;
        public ExpirationResult expiration;
    }
    
    // Global license instance management
    private static volatile TuskLicense instance;
    
    public static TuskLicense initializeLicense(String licenseKey, String apiKey) {
        return initializeLicense(licenseKey, apiKey, null);
    }
    
    public static TuskLicense initializeLicense(String licenseKey, String apiKey, String cacheDir) {
        if (instance == null) {
            synchronized (TuskLicense.class) {
                if (instance == null) {
                    instance = new TuskLicense(licenseKey, apiKey, cacheDir);
                }
            }
        }
        return instance;
    }
    
    public static TuskLicense getLicense() {
        if (instance == null) {
            throw new IllegalStateException("License not initialized. Call initializeLicense() first.");
        }
        return instance;
    }
}

File: sdk/csharp/Tusk.Protection/License.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace Tusk.Protection
{
    /// <summary>
    /// TuskLang SDK License Validation Module
    /// Enterprise-grade license validation for C# SDK
    /// </summary>
    public class TuskLicense
    {
        private readonly string _licenseKey;
        private readonly string _apiKey;
        private readonly string _sessionId;
        private readonly Dictionary<string, LicenseCacheEntry> _licenseCache;
        private readonly List<ValidationAttempt> _validationHistory;
        private readonly List<ExpirationWarning> _expirationWarnings;
        private readonly HttpClient _httpClient;
        private readonly string _cacheDir;
        private readonly string _cacheFile;
        private OfflineCacheData? _offlineCache;
        private readonly ILogger<TuskLicense> _logger;

        public TuskLicense(string licenseKey, string apiKey, string? cacheDir = null, ILogger<TuskLicense>? logger = null)
        {
            _licenseKey = licenseKey;
            _apiKey = apiKey;
            _sessionId = Guid.NewGuid().ToString();
            _licenseCache = new Dictionary<string, LicenseCacheEntry>();
            _validationHistory = new List<ValidationAttempt>();
            _expirationWarnings = new List<ExpirationWarning>();
            _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
            _logger = logger ?? new ConsoleLogger();

            // Set up offline cache directory
            if (cacheDir != null)
            {
                _cacheDir = cacheDir;
            }
            else
            {
                var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                _cacheDir = Path.Combine(homeDir, ".tusk", "license_cache");
            }

            Directory.CreateDirectory(_cacheDir);

            using (var md5 = MD5.Create())
            {
                var hash = md5.ComputeHash(Encoding.UTF8.GetBytes(licenseKey));
                var hashString = BitConverter.ToString(hash).Replace("-", "").ToLower();
                _cacheFile = Path.Combine(_cacheDir, $"{hashString}.cache");
            }

            // Load offline cache if exists
            LoadOfflineCache();
        }

        public ValidationResult ValidateLicenseKey()
        {
            try
            {
                if (string.IsNullOrEmpty(_licenseKey) || _licenseKey.Length < 32)
                {
                    return new ValidationResult { Valid = false, Error = "Invalid license key format" };
                }

                if (!_licenseKey.StartsWith("TUSK-"))
                {
                    return new ValidationResult { Valid = false, Error = "Invalid license key prefix" };
                }

                using (var sha256 = SHA256.Create())
                {
                    var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(_licenseKey));
                    var checksum = BitConverter.ToString(hash).Replace("-", "").ToLower();

                    if (!checksum.StartsWith("tusk"))
                    {
                        return new ValidationResult { Valid = false, Error = "Invalid license key checksum" };
                    }

                    return new ValidationResult { Valid = true, Checksum = checksum };
                }
            }
            catch (Exception ex)
            {
                return new ValidationResult { Valid = false, Error = ex.Message };
            }
        }

        public async Task<Dictionary<string, object>> VerifyLicenseServerAsync(string serverUrl = "https://api.tusklang.org/v1/license")
        {
            try
            {
                var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                var data = new Dictionary<string, object>
                {
                    ["license_key"] = _licenseKey,
                    ["session_id"] = _sessionId,
                    ["timestamp"] = timestamp
                };

                // Generate signature
                var sortedData = JsonSerializer.Serialize(data.OrderBy(kvp => kvp.Key).ToDictionary(kvp => kvp.Key, kvp => kvp.Value));
                using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_apiKey)))
                {
                    var signatureBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(sortedData));
                    data["signature"] = BitConverter.ToString(signatureBytes).Replace("-", "").ToLower();
                }

                var request = new HttpRequestMessage(HttpMethod.Post, serverUrl);
                request.Headers.Add("Authorization", $"Bearer {_apiKey}");
                request.Headers.Add("User-Agent", "TuskLang-CSharp-SDK/1.0.0");
                request.Content = new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json");

                var response = await _httpClient.SendAsync(request);

                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    var result = JsonSerializer.Deserialize<Dictionary<string, object>>(responseContent) ?? new Dictionary<string, object>();

                    // Update in-memory cache
                    _licenseCache[_licenseKey] = new LicenseCacheEntry
                    {
                        Data = result,
                        Timestamp = timestamp,
                        Expires = timestamp + 3600 // Cache for 1 hour
                    };

                    // Update offline cache
                    await SaveOfflineCacheAsync(result);

                    return result;
                }
                else
                {
                    _logger.LogWarning($"Server returned error: {response.StatusCode}");
                    return await FallbackToOfflineCacheAsync($"Server error: {response.StatusCode}");
                }
            }
            catch (HttpRequestException ex)
            {
                _logger.LogWarning($"Network error during license validation: {ex.Message}");
                return await FallbackToOfflineCacheAsync($"Network error: {ex.Message}");
            }
            catch (TaskCanceledException)
            {
                _logger.LogWarning("License validation request timeout");
                return await FallbackToOfflineCacheAsync("Request timeout");
            }
            catch (Exception ex)
            {
                _logger.LogError($"Unexpected error during license validation: {ex.Message}");
                return await FallbackToOfflineCacheAsync(ex.Message);
            }
        }

        public ExpirationResult CheckLicenseExpiration()
        {
            try
            {
                var parts = _licenseKey.Split('-');
                if (parts.Length < 4)
                {
                    return new ExpirationResult { Expired = true, Error = "Invalid license key format" };
                }

                var expirationStr = parts[^1];
                if (!long.TryParse(expirationStr, System.Globalization.NumberStyles.HexNumber, null, out var expirationTimestamp))
                {
                    return new ExpirationResult { Expired = true, Error = "Invalid expiration timestamp" };
                }

                var expirationDate = DateTimeOffset.FromUnixTimeSeconds(expirationTimestamp);
                var currentDate = DateTimeOffset.UtcNow;

                if (expirationDate < currentDate)
                {
                    var daysOverdue = (currentDate - expirationDate).Days;
                    return new ExpirationResult
                    {
                        Expired = true,
                        ExpirationDate = expirationDate.ToString("O"),
                        DaysOverdue = daysOverdue
                    };
                }

                var daysRemaining = (expirationDate - currentDate).Days;

                if (daysRemaining <= 30)
                {
                    _expirationWarnings.Add(new ExpirationWarning
                    {
                        Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                        DaysRemaining = daysRemaining
                    });
                }

                return new ExpirationResult
                {
                    Expired = false,
                    ExpirationDate = expirationDate.ToString("O"),
                    DaysRemaining = daysRemaining,
                    Warning = daysRemaining <= 30
                };
            }
            catch (Exception ex)
            {
                return new ExpirationResult { Expired = true, Error = ex.Message };
            }
        }

        public PermissionResult ValidateLicensePermissions(string feature)
        {
            try
            {
                // Check cached license data
                if (_licenseCache.TryGetValue(_licenseKey, out var cacheEntry))
                {
                    if (DateTimeOffset.UtcNow.ToUnixTimeSeconds() < cacheEntry.Expires)
                    {
                        if (cacheEntry.Data.TryGetValue("features", out var featuresObj) && 
                            featuresObj is JsonElement featuresElement && 
                            featuresElement.ValueKind == JsonValueKind.Array)
                        {
                            var features = featuresElement.EnumerateArray().Select(f => f.GetString()).ToList();
                            if (features.Contains(feature))
                            {
                                return new PermissionResult { Allowed = true, Feature = feature };
                            }
                            else
                            {
                                return new PermissionResult { Allowed = false, Feature = feature, Error = "Feature not licensed" };
                            }
                        }
                    }
                }

                // Fallback to basic validation
                switch (feature)
                {
                    case "basic":
                    case "core":
                    case "standard":
                        return new PermissionResult { Allowed = true, Feature = feature };
                    case "premium":
                    case "enterprise":
                        if (_licenseKey.ToUpper().Contains("PREMIUM") || _licenseKey.ToUpper().Contains("ENTERPRISE"))
                        {
                            return new PermissionResult { Allowed = true, Feature = feature };
                        }
                        else
                        {
                            return new PermissionResult { Allowed = false, Feature = feature, Error = "Premium license required" };
                        }
                    default:
                        return new PermissionResult { Allowed = false, Feature = feature, Error = "Unknown feature" };
                }
            }
            catch (Exception ex)
            {
                return new PermissionResult { Allowed = false, Feature = feature, Error = ex.Message };
            }
        }

        public LicenseInfo GetLicenseInfo()
        {
            var validationResult = ValidateLicenseKey();
            var expirationResult = CheckLicenseExpiration();

            var info = new LicenseInfo
            {
                LicenseKey = $"{_licenseKey.Substring(0, Math.Min(8, _licenseKey.Length))}...{_licenseKey.Substring(Math.Max(0, _licenseKey.Length - 4))}",
                SessionId = _sessionId,
                Validation = validationResult,
                Expiration = expirationResult,
                CacheStatus = _licenseCache.ContainsKey(_licenseKey) ? "cached" : "not_cached",
                ValidationCount = _validationHistory.Count,
                Warnings = _expirationWarnings.Count
            };

            if (_licenseCache.TryGetValue(_licenseKey, out var cacheEntry))
            {
                info.CachedData = cacheEntry.Data;
                info.CacheAge = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - cacheEntry.Timestamp;
            }

            return info;
        }

        public void LogValidationAttempt(bool success, string details = "")
        {
            _validationHistory.Add(new ValidationAttempt
            {
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                Success = success,
                Details = details,
                SessionId = _sessionId
            });
        }

        private void LoadOfflineCache()
        {
            try
            {
                if (File.Exists(_cacheFile))
                {
                    var json = File.ReadAllText(_cacheFile);
                    var cached = JsonSerializer.Deserialize<OfflineCacheData>(json);

                    if (cached != null)
                    {
                        // Verify the cache is for the correct license key
                        using (var sha256 = SHA256.Create())
                        {
                            var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(_licenseKey));
                            var keyHash = BitConverter.ToString(hash).Replace("-", "").ToLower();

                            if (cached.LicenseKeyHash == keyHash)
                            {
                                _offlineCache = cached;
                                _logger.LogInformation("Loaded offline license cache");
                            }
                            else
                            {
                                _logger.LogWarning("Offline cache key mismatch");
                                _offlineCache = null;
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogError($"Failed to load offline cache: {ex.Message}");
                _offlineCache = null;
            }
        }

        private async Task SaveOfflineCacheAsync(Dictionary<string, object> licenseData)
        {
            try
            {
                using (var sha256 = SHA256.Create())
                {
                    var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(_licenseKey));
                    var keyHash = BitConverter.ToString(hash).Replace("-", "").ToLower();

                    var cacheData = new OfflineCacheData
                    {
                        LicenseKeyHash = keyHash,
                        LicenseData = licenseData,
                        Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                        Expiration = CheckLicenseExpiration()
                    };

                    var json = JsonSerializer.Serialize(cacheData, new JsonSerializerOptions { WriteIndented = true });
                    await File.WriteAllTextAsync(_cacheFile, json);
                    _offlineCache = cacheData;
                    _logger.LogInformation("Saved license data to offline cache");
                }
            }
            catch (Exception ex)
            {
                _logger.LogError($"Failed to save offline cache: {ex.Message}");
            }
        }

        private async Task<Dictionary<string, object>> FallbackToOfflineCacheAsync(string errorMsg)
        {
            if (_offlineCache != null && _offlineCache.LicenseData != null)
            {
                var cacheAge = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - _offlineCache.Timestamp;
                var cacheAgeDays = cacheAge / 86400.0;

                // Check if cached license is not expired
                if (!_offlineCache.Expiration.Expired)
                {
                    _logger.LogWarning($"Using offline license cache (age: {cacheAgeDays:F1} days)");
                    var result = new Dictionary<string, object>(_offlineCache.LicenseData)
                    {
                        ["offline_mode"] = true,
                        ["cache_age_days"] = cacheAgeDays,
                        ["warning"] = $"Operating in offline mode due to: {errorMsg}"
                    };
                    return result;
                }
                else
                {
                    return new Dictionary<string, object>
                    {
                        ["valid"] = false,
                        ["error"] = $"License expired and server unreachable: {errorMsg}",
                        ["offline_cache_expired"] = true
                    };
                }
            }
            else
            {
                return new Dictionary<string, object>
                {
                    ["valid"] = false,
                    ["error"] = $"No offline cache available: {errorMsg}",
                    ["offline_cache_missing"] = true
                };
            }
        }

        // Data classes
        public class ValidationResult
        {
            public bool Valid { get; set; }
            public string? Error { get; set; }
            public string? Checksum { get; set; }
        }

        public class ExpirationResult
        {
            public bool Expired { get; set; }
            public string? Error { get; set; }
            public string? ExpirationDate { get; set; }
            public long? DaysRemaining { get; set; }
            public long? DaysOverdue { get; set; }
            public bool? Warning { get; set; }
        }

        public class PermissionResult
        {
            public bool Allowed { get; set; }
            public string Feature { get; set; } = "";
            public string? Error { get; set; }
        }

        public class LicenseInfo
        {
            public string LicenseKey { get; set; } = "";
            public string SessionId { get; set; } = "";
            public ValidationResult Validation { get; set; } = new();
            public ExpirationResult Expiration { get; set; } = new();
            public string CacheStatus { get; set; } = "";
            public int ValidationCount { get; set; }
            public int Warnings { get; set; }
            public Dictionary<string, object>? CachedData { get; set; }
            public long? CacheAge { get; set; }
        }

        private class LicenseCacheEntry
        {
            public Dictionary<string, object> Data { get; set; } = new();
            public long Timestamp { get; set; }
            public long Expires { get; set; }
        }

        private class ValidationAttempt
        {
            public long Timestamp { get; set; }
            public bool Success { get; set; }
            public string Details { get; set; } = "";
            public string SessionId { get; set; } = "";
        }

        private class ExpirationWarning
        {
            public long Timestamp { get; set; }
            public long DaysRemaining { get; set; }
        }

        private class OfflineCacheData
        {
            public string LicenseKeyHash { get; set; } = "";
            public Dictionary<string, object> LicenseData { get; set; } = new();
            public long Timestamp { get; set; }
            public ExpirationResult Expiration { get; set; } = new();
        }

        // Simple console logger implementation
        private class ConsoleLogger : ILogger<TuskLicense>
        {
            public IDisposable BeginScope<TState>(TState state) => new NoopDisposable();
            public bool IsEnabled(LogLevel logLevel) => true;
            public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
            {
                Console.WriteLine($"[TuskLicense] [{logLevel}] {formatter(state, exception)}");
            }

            private class NoopDisposable : IDisposable
            {
                public void Dispose() { }
            }
        }
    }

    // Global license instance management
    public static class TuskLicenseManager
    {
        private static TuskLicense? _instance;
        private static readonly object _lock = new object();

        public static TuskLicense InitializeLicense(string licenseKey, string apiKey, string? cacheDir = null, ILogger<TuskLicense>? logger = null)
        {
            lock (_lock)
            {
                _instance = new TuskLicense(licenseKey, apiKey, cacheDir, logger);
                return _instance;
            }
        }

        public static TuskLicense GetLicense()
        {
            lock (_lock)
            {
                if (_instance == null)
                {
                    throw new InvalidOperationException("License not initialized. Call InitializeLicense() first.");
                }
                return _instance;
            }
        }
    }
}